core: add a KEP for modeling entitlements to cross-workspace concerns (#4)
kcp-dev/enhancements: keps/core/model-cross-worksapce-entitlements.md

  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
<!-- /toc -->

## Summary

This KEP describes an API that would allow users of `kcp` to predicate use of
their cross-workspace relationships on an entitlement system. Furthermore, this
KEP introduces a provisional implementation of the API that will suffice for
bootstrapping and small-scale environments.

## Motivation

`kcp` distinguishes itself from other Kubernetes-based control planes by providing
not only cheap multi-tenancy through workspaces but also allowing users to model
cross-workspace concerns. While a fleet of entirely independent workspaces has
value, it's these cross-workspace interactions that are uniquely possible in `kcp`.
One common type of cross-workspace interaction is a provider/consumer split - one

> actor or set of actors provide some service through content in their workspace, and

Reviewer: Would remove "or a set of actors"

others using the system can consume that service. Examples of this relationship include:

 - `APIExports`, used via `APIBindings` from the `apis.kcp.io` group
 - `WorkspaceTypes`, used in creating `Workspaces` from the `tenancy.kcp.io` group
 - `Locations`, used via `Placements` from the `scheduling.kcp.io` group

Naturally, providers of these functionalities would like to manage where their services
can be used. Today, restricting access to the provider's service is done in an ad-hoc

> manner (or not implementded at all!), for example with `APIExports` the flow looks like:

Reviewer: ```suggestion
manner (or not implemented at all!); for example, with `APIExports` the flow looks like:
```

 - create RBAC in the `APIExport` workspace that provides `verb="bind"` access to the
   particular `APIExport` name in question
 - gate access to the cross-workspace API by implementing a validating admission plugin
   that issues a `SubjectAccessReview` using a "deep" client that skips other authorizers
   while issuing the SAR

Notably, use of the "deep" `SubjectAccessReview` client requires an admission _plugin_,
and is privileged. Therefore, the bar for implementing new cross-workspace concepts
with any sort of authorization is high. Even for in-tree, privileged APIs, this is a
non-trivial implementation cost.

This KEP proposes a mechanism to model entitlements to cross-workspace concepts both for
in-tree uses as well as user-provided extensions.

### Goals

1. Model entitlements to cross-workspace concerns in a generic manner
1. Enable entitlement checking in a non-privileged manner
1. Provide an implementation for entitlement checking that suffices for bootstrapping

> 1. Simplify the `kcp` authorizer chain by removing non-Kubernetes, non-RBAC concepts

Reviewer 2: + remove deep SAR ?

### Non-Goals

1. Enforce a particular data model or implementation

> 1. Provide an implementation that scales well

Reviewer 2: is "allow a scalable implementation" a goal?

1. Remove all custom RBAC objects from `kcp` systems

## Proposal

`kcp` will expose a new API for reviewing entitlements, the implementation of which
will be configured at server start-up. This endpoint will act as a well-known proxy
to the implementation being used, in order to allow downstream consumers to be
unaware of changes to the implementation. Access to this API will be done through
workspace-specific endpoints and authorized on a per-endpoint basis. Users building
cross-workspace concepts on top of `kcp` will be able to write admission _webhooks_
that use this API to implement restrictions to the use of their concepts.

Additionally, a default implemenation will be provided for `kcp start` to allow for

> boot-strappiung the system. This implemenation will be self-sufficient at small scales,

Reviewer: ```suggestion
bootstrapping the system. This implemenation will be self-sufficient at small scales,
```

For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

A new endpoint will be exposed to facilitate entitlement review. Users will `POST`
an `EntitlementReview` object to the endpoint and synchronously receive the result
in the response. The `EntitlementReview` has the following form:

```go
type EntitlementReview struct {
    metav1.TypeMeta `json:",inline"`
    // Spec holds information about the request being evaluated.
    Spec EntitlementReviewSpec `json:"spec"`

    // Status is filled in by the server and indicates whether the request is allowed or not.
    // +optional
    Status EntitlementReviewStatus `json:"status,omitempty"`
}

// EntitlementReviewSpec is a description of the entitlement request.
type EntitlementReviewSpec struct {
    // UserInfo is information about the requesting user
    UserInfo authenticationv1.UserInfo `json:"userInfo"`

    // RequestInfo is information about the context in which this request was issued.

>     RequestInfo RequestInfo `json:"requestInfo"`

Reviewer 2: aren't these implicit in the request handling of the apiserver? No need to have them in the object posted by the client.

Author: There are two hops - client to API server, where these *are* implicit, and admission webhook to entitlement reviewer, where they are *not*.

Reviewer 2: ack

    ClusterPath logicalcluster.Path `json:"clusterPath"`
}

// Entitlement describes an entitlement
type Entitlement struct {
    // ClusterName is the name of the logical cluster from which the entitlement is provided.
    // NOTE: the entitlement review API is segmented by this cluster name in order to allow
    // fine-grained authorization for the review API, so this field is superfluous; however, 
    // duplicating the data here allows for the entitlement review object to be self-sufficient.
    // +optional
    ClusterName logicalcluster.Name `json:"clusterName"`

    // The type of entitlement being described below is free-form data to be interpreted by
    // the implementor of the review server. This type metadata allows for the object to be
    // implemented as a discriminated union.
    metav1.TypeMeta `json:",inline"`

    // Spec is the free-form entitlement to be reviewed.
    Spec runtime.RawExtension `json:"spec,omitempty"`
}

// EntitlementReviewStatus contains the result of the entitlement review.
type EntitlementReviewStatus struct {
    // Entitled is required. True if the entitlement is granted, false otherwise.
    Entitled bool `json:"entitled"`

>     // Disqualified is optional. True if the entitlement  is not granted, otherwise

Reviewer 2: ```suggestion
    // Disqualified is optional. True if the entitlement is not granted, otherwise
```

    // +optional
    Reason string `json:"reason,omitempty"`
    // EvaluationError is an indication that some error occurred during the entitlement review.
    // It is entirely possible to get an error and be able to continue determine entitlement status in spite of it.
    // +optional
    EvaluationError string `json:"evaluationError,omitempty"`
}
```

The entitlement review API will be partitioned by workspace; entitlement reviews will be issued
to the workspace-specific endpoint relevant to the entitlement and access to this API will be 
guarded by Kubernetes RBAC in the workspace:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: core:entitlements:review
rules:
- apiGroups: ["core.kcp.io"]
  resources:
  - "entitlementreviews"
  verbs: ["create"]
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: core:entitlements:review
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: core:entitlements:review
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:authenticated
```

Use of the entitlement review API might look like

```

> > POST /services/entitlementreview/clusters/<entitlement-provider-cluster>/apis/core.kcp.io/v1alpha1/entitlementreviews

Reviewer 2: this surprises. Would have expected to have the review as a local request, not against the provider cluster. Also having this under `/services` is odd.

Reviewer 2: This is related to our private discussion in Slack about inversion of the mechanism:
- make entitlements properties of the acting workspace.
- allow a provider workspace to control entitlements.

Author: It's against the (local) view server, which is as local as we can get, right? It's not "against" the provider cluster, but we need to indicate somehow that we're reviewing an entitlement _from_ somewhere.

Reviewer 2: why not `/clusters/<workspace-checked-for-entitlement>/...`?

Reviewer 2: Commented somewhere else already: every service URL must be announced somewhere in another API object the caller has access to. `/services` as a prefix is not a convention and can vary by kcp installation and be completely different or on another host.

    // Entitlements holds all the entitlements that this EntitlementPolicy
    // groups. 
    // +optional
    Entitlements []kcpcorev1alpha1.Entitlement `json:"entitlements"`
}

type EntitlementPolicyBinding struct {
    metav1.TypeMeta `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // EntitlementPolicyRef is a reference to the entitlement policy
    // that this binding applies to.
    EntitlementPolicyRef EntitlementPolicyRef `json:"entitlementPolicyRef"`

    // Children determines if this policy binding provides the entitlement just to the
    // workspace in which it is created, or also to the child workspaces underneath that.
    // Optional, defaults to false.
    // +optional
    Children boolean `json:"children,omitempty"`
}

type EntitlementPolicyRef struct {
    // ClusterPath is the human-readable path to the logical cluster in which the entitlement
    // policy is registered.
    ClusterPath logicalcluster.Path `json:"clusterPath"`

    // Name is the name of the entitlement policy.
    Name string `json:"name"`
}
```

A service provider with permissions to create `EntitlementPolicy` will do so using the
normal `kcp` API:

```

> > POST /clusters/<entitlement-provider-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicies

Reviewer 2: as above, IMO the wrong direction. Put these into the entitled workspace. The current design here means that number of policies grow with number of entitled workspaces. E.g. this means for every home workspace you need this object too. IMO a data model that does not fit the scaling dimensions.

Reviewer 2: Oh, further down:

> Entitlement policy bindings will be stored in tenant clusters

Maybe just a typo here? `s/<entitlement-provider-cluster>/<tenant-cluster>/` ?

Reviewer: From what I think I'm reading, what Steve has in the doc is what he intended. The service provider == the entitlement provider cluster. The EntitlementPolicy with a list of Entitlement(s) is like an RBAC Role. Then, in the consumer cluster, there is an EntitlementPolicyBinding that references the EntitlementPolicy in the service provider cluster. This is similar to a RoleBinding. Am I reading this right, @stevekuznetsov? If so, the only question I think I have is why the entitlement in the policy has a clusterName?

Author: @ncdc yep, that's what I intended. As for why an `entitlements.core.kcp.io` has a cluster name ... I went back and forth on this. It's always possible to infer the cluster name from an entitlement review check based on the path that someone used to issue it, but I didn't know if that would be annoying / make implementing a replacement review server more irritating than it needed to be.

Reviewer: Is the clusterName in the entitlement in the policy the logical cluster where the service (e.g. APIExport) lives?

Author: Yes! Also, the current design requires that the service (e.g. APIExport) and entitlement share a logical cluster.

> {
>     "kind": "EntitlementPolicy",
>     "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
>     "metadata": {
>         "name": "something-meaningful"
>     },
>     "entitlements": [
>         {
>             "clusterName": "33bab531",
>             "kind": "WidgetAccess",
>             "apiVersion": "acme.corp/v1",
>             "spec": {
>                 "something": {
>                     "meaningful": "here"
>                 }
>             }
>         }
>     ]
> }
201 CREATED
{
    "kind": "EntitlementPolicy",
    "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
    "metadata": {
        "name": "something-meaningful"
    },
    "entitlements": [
        {
            "clusterName": "33bab531",
            "kind": "WidgetAccess",
            "apiVersion": "acme.corp/v1",
            "spec": {
                "something": {
                    "meaningful": "here"
                }
            }
        }
    ]
}
```

However, to entitle tenants, a per-tenant endpoint is used:

```

> > POST /services/entitlements/clusters/<entitlement-consumer-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicybindings

Reviewer 2: maybe intentional: this means that this POST goes to the same shard as the APIExport workspace? And that the VW is clever enough and has permissions to route the request to the right shard where the tenant workspace lives? This might be possible to do, but need deeper design.

Also: where does an entitler know this URL from? It must be set somewhere, e.g. on an APIExport.

Alternative might be that we have an `ThingEntitlementClass` object in the provider workspace, and we put the URL there into status.

Author: > maybe intentional: this means that this POST goes to the same shard as the APIExport workspace?

I expect this to be unrelated from APIExports, actually - we already have three distinct use-cases in our current code-base and I imagine Edge MC will have ones as well. 

> And that the VW is clever enough and has permissions to route the request to the right shard where the tenant workspace lives? This might be possible to do, but need deeper design.

Orthogonally from that, though, yes, if we expect a user to interact with one endpoint and we need the output objects to go to the right shard, we need that routing logic. Doesn't that problem exist already today with any other view?

> Also: where does an entitler know this URL from? It must be set somewhere, e.g. on an APIExport.

I was expecting something like your `ThingEntitlementClass` idea, good call-out that this is underspecified.

Reviewer 2: > Doesn't that problem exist already today with any other view?

Am not aware of any such case. It's not trivial I fear, basically depends on the workspace index.

Reviewer 2: One thought: would add the provider cluster and policy name in here:

```
POST /services/entitlements/clusters/<provider>/<policy-name>/clusters/<entitlement-consumer-cluster>/apis/entitlements.tenancy.kcp.io/v1alpha1/entitlementpolicybindings
```

That way we could authz the access easily against the policy object.

> {
>     "kind": "EntitlementPolicyBinding",
>     "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
>     "metadata": {
>         "name": "something-meaningful-for-a-tenant"
>     },
>     "entitlementPolicyRef": {
>         "clusterPath": "<entitlement-provider-cluster>",
>         "name": "something-meaningful"
>     },
>     children: false
> }
201 CREATED
{
    "kind": "EntitlementPolicyBinding",
    "apiVersion": "entitlements.tenancy.kcp.io/v1alpha1",
    "metadata": {
        "name": "something-meaningful-for-a-tenant"
    },
    "entitlementPolicyRef": {
        "clusterPath": "<entitlement-provider-cluster>",
        "name": "something-meaningful"
    },
    children: false
}
```

Entitlement policy bindings will be stored in tenant clusters; however, the tenant will have

> no access to read or change them. This mechanism is used to distribute this data, reduce its

Reviewer 2: why not read?

Author: I was hoping to strike a middle ground between providing a useful base implementation while not exposing the APIs anywhere they don't need to be exposed, so it's less likely that when a deployment of `kcp` changes the implementation mechanism, people have their workflows broken. Perhaps useful to sketch out what it looks like when a custom entitlement review is added, though - no real reason to remove this one from the system at that time, it could still function, as a fallback?

Reviewer 2: I like that thought :+1: If you cannot read them, nobody will ask about listing either, and with that nobody will ask for an exhaustive list of entitlement including external ones.

cardinality and remove the need to cache policy bindings that do not implicate children. For
those entitlement policy bindings which _do_ set `"children":true`, replication using the cache
server will be required.

> A server will also exists that can field entitlement review requests. This server will entirely

Reviewer: ```suggestion
A server will also exist that can field entitlement review requests. This server will entirely
```
